Add contextual options to expression-attributes design doc (#780)
unicode-org/message-format-wg: exploration/u-options.md

> # `u:` Options

Author: Maybe a bit shorter even?
```suggestion
# Contextual Options
```

> Status: **Accepted**

Author: I would prefer starting with Accepted, and then only if necessary first merging it as Proposed. We should not start from the presumption that multiple phases of discussion are required for everything we consider via design documents.

> Status: **Accepted**

Reviewer: We're going to discuss it in an hour anyway 😄

<details>
	<summary>Metadata</summary>
	<dl>
		<dt>Contributors</dt>
		<dd>@eemeli</dd>
		<dt>First proposed</dt>
		<dd>2023-05-06</dd>
		<dt>Pull Requests</dt>
		<dd><a href="https://github.com/unicode-org/message-format-wg/pull/780">#780</a></dd>
	</dl>
</details>

## Objective

Provide common definitions for contextual options.

## Background

Function options may influence the resolution, selection, and formatting of annotated expressions.
These provide a great solution for options like `minFractionDigits`, `dateStyle`,
or other similar factors that influence the formatted result.

> However, this single bag of options is not appropriate in all cases,
> in particular for options that change the formatting context for the function.

Reviewer: I'm not sure this is what you mean. Perhaps:

```suggestion
Such options naturally correspond to function arguments or builder-style function constructors.
Each option is specific to the associated API.
Users (message authors, translators, developers) should not expect
that a given option name has the same meaning between functions
or that its behavior stays the same.

To reduce the learning curve for users and improve consistency,
it would be useful to have common options 
(generally those related to the formatting context)
shared between all functions.
```

## Use-Cases

At least the following function and markup options should be considered:

- An identifier for the expression or markup.
  This is included in the formatted part,
  and allows the parts of an expression to be explicitly addressed.

  > Example identifying two literal numbers:
  >
  > ```
  > The first number was {1234 :number u:id=first} and the second {56789 :number u:id=second}.
  > ```

- An override for the locale used to format the expression.

>   Should be expressed as a non-empty sequence of BCP 47 language codes.

Reviewer: Sequence?

```suggestion
The value of this option is a Unicode Locale Identifier
(which is a BCP47 language tag).
The root locale is expressed using the tag `und`.
```

Author: I copied this from the expression-attributes text. As this is describing a use case, it's perhaps best to not get stuck in the technical details here.
```suggestion
```

>   > Example embedding a French literal in an English message:
>   >
>   > ```
>   > In French, “{|bonjour| :string u:locale=fr}” is a greeting
>   > ```

Reviewer: This isn't perhaps a great example.

```suggestion
> Example. Overriding the formatting locale to use French:
>```
> In French, this date would be displayed as {|2024-05-06| :date u:locale=fr}
>```
```

Author: ```suggestion
  > Example embedding a French date in an English message:
  >
  > ```
  > In French, this date would be displayed as {|2024-05-06| :date u:locale=fr}
  > ```
```
